File: Swift/Leetcode.playground/Sources/14_Graph_DFS/LC_2698_FindThePunishmentNumberOfAnInteger.swift
class Solution {
    func punishmentNumber(_ n: Int) -> Int {
        var ans = 0
        for i in 1...n {
            if dfs(i * i, i) {
                ans += i * i
            }
        }
        return ans
    }
    
    func dfs(_ t: Int, _ x: Int) -> Bool {
        if t == x {
            return true
        }
        var d = 10
        while t >= d && t % d <= x {
            if dfs(t / d, x - (t % d)) {
                return true
            }
            d *= 10
        }
        return false
    }
}

File: Java/14_Graph_DFS/LC_2698_FindThePunishmentNumberOfAnInteger.java
class Solution {
    public int punishmentNumber(int n) {
        int ans = 0;
        for (int i = 1; i <= n; i++) {
            if (dfs(i * i, i))
                ans += i * i;
        }
        return ans;
    }

    boolean dfs(int t, int x) {
        if (t == x)
            return true;
        int d = 10;
        while (t >= d && t % d <= x) {
            if (dfs(t / d, x - (t % d)))
                return true;
            d *= 10;
        }
        return false;
    }
}

public class LC_2698_FindThePunishmentNumberOfAnInteger {

}
